### ISSUE TITLE
Improve ArgoCD controller CPU usage by using marshmallow to do marshall/unmarshall in certain places.

### ISSUE THREAD

**Issue by wilsonwang371** [argoproj/argo-cd#17130]
# Summary

There are tons of Marshalling/Unmarshalling operations in ArgoCD and some of them are at the critical paths which can affect overall performance. We want to use marshmallow library to parse the data instead of using the unstructured marshalling/unmarshalling which can take significant more CPU usages.

# Motivation

We found that we reduced peak CPU usage by 30% in our downstream argocd by using marshmallow in some of the controller code path.

# Proposal

In getResourceTree(), we can use marshmallow to speed up PartialJson parsing. The reason is that we just use only a few fields of the object which does not need to be fully parsed.

**Comment by crenshaw-dev** [comment]
This is super interesting. Do you have a draft PR you can share and any additional stats you have about the controllers behavior with/without the patch?

**Comment by wilsonwang371** [comment]
My patch is here https://github.com/wilsonwang371/argo-cd/commit/923fc12ee86accdc2e48390a83afd85045ad98b5

> This is super interesting. Do you have a draft PR you can share and any additional stats you have about the controllers behavior with/without the patch?

**Comment by wilsonwang371** [comment]
![image](https://github.com/argoproj/argo-cd/assets/3913185/d53f61e6-c89b-49bb-8757-be02e46d729d)
This is the image showing CPU usage reduction in our controller cluster. I don't have more detailed information. Maybe we can run this patch again and collect further data.

**Comment by crenshaw-dev** [comment]
I'm also curious if you encountered any other "hot paths" for unmarshalling which we should keep in mind for future optimization?

**Comment by wilsonwang371** [comment]
> I'm also curious if you encountered any other "hot paths" for unmarshalling which we should keep in mind for future optimization?

right now, the one that I patches is the one that we can get most of the performance improvement if fixed. There are other places that we can do similar things, but so far they won't give us enough performance improvement as this one does.

**Comment by wilsonwang371** [comment]
> Nice. So we're unmarshalling fewer fields, makes sense that it would be faster. Is the switch to the marshmallow lib necessary, or would the built in lib perform similarly on the smaller field set?

Here is the link to marshmallow's details. https://docs.google.com/presentation/d/1sFAY2JZnCRA3t_VzZggyr9qa5A2sYHSxaKH6phVECJE/edit#slide=id.p46

Ideally, we can use go codec to get similar performance. However, marshmallow claimed that they did more optimizations such as caching which make their implementation performance better. So IMHO, directly using marshmallow makes the code simpler.

**Comment by woehrl01** [comment]
@wilsonwang371 this looks very interesting. Did you alps made some evaluations with https://github.com/valyala/fastjson?

I don't have used marshmellow yet, only fastjson. If I look at your PR it looks like that using fastjson could add additional performance benefits, by having the possibility of caching the parser. As you only need the same "general" fields of a CRD.

**Comment by wilsonwang371** [comment]
> @wilsonwang371 this looks very interesting. Did you also made some evaluations with https://github.com/valyala/fastjson?
> 
> I don't have used marshmellow yet, only fastjson. If I look at your PR it looks like that using fastjson could add additional performance benefits, by having the possibility of caching the parser and no need of using reflect, as you only need the same "general" fields of a CRD. This should also have the added benefit of having mostly zero additional allocation which reduces the pressure on the gc.

Thanks for the info, I am right now working on a solution which will not require a full json object parse. In this way, it will be way faster than most of the available solutions.

**Comment by wilsonwang371** [comment]
with partial json parsing, to parse a normal argocd application json format. My code is 30.333µs compared with fastjson 62.542µs and encoding/json takes around 145.291µs. I will post a patch later.

**Comment by andklee** [comment]
@wilsonwang371 Spoke about this during the sig-scalability call, wondering if we could do some benchmarking on this change to test the different libraries (marshmallow vs encoding/json) on a standardized argo-cd performance test (Perhaps CPU/Memory of the argo-cd controller). I can perhaps run this on my environment with 100 clusters/10000 apps. If you have details about your changes. let me know.

**Comment by wilsonwang371** [comment]
> @wilsonwang371 Spoke about this during the sig-scalability call, wondering if we could do some benchmarking on this change to test the different libraries (marshmallow vs encoding/json) on a standardized argo-cd performance test (Perhaps CPU/Memory of the argo-cd controller). I can perhaps run this on my environment with 100 clusters/10000 apps. If you have details about your changes. let me know.

Sounds good. What kind of benchmarking shall we do? Simply testing the performance by running the function or measure them in a real environment?

**Comment by wilsonwang371** [comment]
> @wilsonwang371 Spoke about this during the sig-scalability call, wondering if we could do some benchmarking on this change to test the different libraries (marshmallow vs encoding/json) on a standardized argo-cd performance test (Perhaps CPU/Memory of the argo-cd controller). I can perhaps run this on my environment with 100 clusters/10000 apps. If you have details about your changes. let me know.

The code is now at #17802

**Comment by Enclavet** [comment]
> > @wilsonwang371 Spoke about this during the sig-scalability call, wondering if we could do some benchmarking on this change to test the different libraries (marshmallow vs encoding/json) on a standardized argo-cd performance test (Perhaps CPU/Memory of the argo-cd controller). I can perhaps run this on my environment with 100 clusters/10000 apps. If you have details about your changes. let me know.
> 
> Sounds good. What kind of benchmarking shall we do? Simply testing the performance by running the function or measure them in a real environment?

We would run them in a real environment. I noticed from your PR you are not using marshmellow anymore? So looks like you have fastjson and ujson (your own implementation to do a partial unmarshall?).

**Comment by wilsonwang371** [comment]
> > > @wilsonwang371 Spoke about this during the sig-scalability call, wondering if we could do some benchmarking on this change to test the different libraries (marshmallow vs encoding/json) on a standardized argo-cd performance test (Perhaps CPU/Memory of the argo-cd controller). I can perhaps run this on my environment with 100 clusters/10000 apps. If you have details about your changes. let me know.
> > 
> > 
> > Sounds good. What kind of benchmarking shall we do? Simply testing the performance by running the function or measure them in a real environment?
> 
> We would run them in a real environment. I noticed from your PR you are not using marshmellow anymore? So looks like you have fastjson and ujson (your own implementation to do a partial unmarshall?).

Correct. Since ujson is way faster than marshmallow, I am just comparing fastjson and ujson. BTW, there are some minor issues, I am fixing it now

**Comment by wilsonwang371** [comment]
> > > @wilsonwang371 Spoke about this during the sig-scalability call, wondering if we could do some benchmarking on this change to test the different libraries (marshmallow vs encoding/json) on a standardized argo-cd performance test (Perhaps CPU/Memory of the argo-cd controller). I can perhaps run this on my environment with 100 clusters/10000 apps. If you have details about your changes. let me know.
> > 
> > 
> > Sounds good. What kind of benchmarking shall we do? Simply testing the performance by running the function or measure them in a real environment?
> 
> We would run them in a real environment. I noticed from your PR you are not using marshmellow anymore? So looks like you have fastjson and ujson (your own implementation to do a partial unmarshall?).

You can try the latest MR code. It should be working.

**Comment by Enclavet** [comment]
Perfect. I see it now and I will try it out. Will perform my testing and post my results here when I find some time.
Probably will compare the original code vs fastjson and ujson.

**Comment by wilsonwang371** [comment]
> Perfect. I see it now and I will try it out. Will perform my testing and post my results here when I find some time. Probably will compare the original code vs fastjson and ujson.

thanks, BTW, we have a bunch of downstream performance improvement patches. These patches significantly improved our argocd performance. I haven't posted them yet. They are already in our production environment. I will gradually post these changes later.

**Comment by wilsonwang371** [comment]
> Perfect. I see it now and I will try it out. Will perform my testing and post my results here when I find some time. Probably will compare the original code vs fastjson and ujson.

I already have some tests that compared the result. It is already inside the code.
